clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -56,87 +59,108 @@ WildFly needs an persistent timer service implementation suitable for use in a c
+
+The general implementation plan for this feature is as follows:
+
+* Introduce a proper SPI for a `TimerService` within `org.jboss.as.ejb3.timerservice.spi`, to facilitate multiple implementations.
+** Adapt the existing `TimerResource` runtime attributes/operations to use the `TimerService` SPI instead of referencing specific implementation classes.
+** Adapt existing `TimerService` implementation to this SPI.
+* Create `CompositeTimerService`, a `TimerService` implementation that delegates to separate persistent and transient `TimerService` implementations.
+* Implement an distributed implementation of the `TimerService` SPI.
+** Add a `org.wildfly.clustering.ejb.timer` SPI to the `wildfly-clustering-ejb-spi` module to facilitate multiple distributed timer manager implementations.
+** Create `DistributedTimerService`, a distributed `TimerService` implementation to the ejb3 subsystem that delegates to the `wildfly-clustering-ejb-spi` for timer management and scheduling.
+** Add an Infinispan implementation of the `org.wildfly.clustering.ejb.timer` SPI that assigns execution ownership of timers to cluster members based on consistent hashing of the timer identifier.
+*** Leverage the existing `LocalScheduler` implementation within `wildfly-clustering-ee-cache` for timer scheduling
+*** Leverage the existing `PrimaryOwnerScheduler` decorator that uses consistent hashing of the timer identifier to distribute timer execution among cluster members
+* Add a management resource and capability to the distributable-ejb subsystem to configure the Infinispan timer management implementation.
+* Make the `default-data-store` attribute of the timer service resource optional.
+** Require the default distributable timer management capability when undefined.
+** Modify the `TimerService` DUP to use a `CompositeTimerService` in the event that the server is configured to use persistent timer management, that delegates to the existing `TimerServiceImpl` for use with transient timers and the `DistributableTimerService` for use with persistent timers.
 
 == Security Considerations
 
-////
-Identification if any security implications that may need to be considered with this feature
-or a confirmation that there are no security implications to consider.
-////
+N/A
 
 == Test Plan
 
-* All existing TimerService tests must pass against the new persistent timer implementation, including the EJB TCK.
-* Existing distributed EJB timer tests that use the JDBC timer-store within an EAP cluster may need to be modified since timers may execute on different cluster members than expected by the tests.
-* New tests will be added to the clustering integration testsuite to validate correctness of behavior with a cluster, e.g. failover behavior, etc.
+* All existing `TimerService` tests must pass against the new persistent timer implementation, including the Jakarta Enterprise Beans TCK.
```

Comment:
Is the Jakarta Enterprise Beans TCK execution bound to any WildFly CI execution, like e.g. it happens for the Microprofile TCKs? Basically, in which context/domain will these tests be executed?

Reply:
The execution of the Jakarta EE TCK is completely external to WildFly.  @scottmarlow can answer any specific questions you have about this.

Reply:
Thanks @pferraro - Hi @scottmarlow, my point here is to understand which environment/domain/context should be checked in oder to verify the execution of this TCK, when testing the feature.

Reply:
@fabiobrz See the test result attachment in the downstream JIRA.

Reply:
> Thanks @pferraro - Hi @scottmarlow, my point here is to understand which environment/domain/context should be checked in oder to verify the execution of this TCK, when testing the feature.

Hi @fabiobrz, the TCK testing is internal to Red Hat currently.

Reply:
Thanks for your feedback on this @pferraro and @scottmarlow - I'll check the resources.